### DISCUSSION TITLE
What profilers do you all use?

### DISCUSSION THREAD

**Discussion by LowLevelMahn** [dosbox-staging/dosbox-staging#1261] [General]
i got very good results with Intel VTune (wich is free and also available for Linux, but only useable on Intel Hardware) helping me to optimize some network software very much

what are your standard tools for profiling?

**Comment by kcgen** [comment]
@kklobe has a nice setup on the Apple M1.  Learning about Apple's profiling tools could pay dividends for all platforms.

**Comment by kcgen** [comment]
I wrote a little about using the Intel compiler here: https://github.com/dosbox-staging/dosbox-staging/wiki/Intel-Compiler-Tips

Any help or tips you can give on VTune will be very welcome! (tutorials, etc.. I can add them to that wiki page).

**Comment by kcgen** [comment]
We've used the Linux kernel's flamegraphs to show cumulative time spent in nested functions: https://github.com/dosbox-staging/dosbox-staging/pull/887 (later in that PR I added manual `std::chrono` measurements to measure the latency experienced by the mixer when waiting for samples from the device).

**Comment by kklobe** [comment]
> @kklobe has a nice setup on the Apple M1. Learning about Apple's profiling tools could pay dividends for all platforms.

One thing that's great about the Apple tools is that there's zero friction. No special recompiling to instrument the binary, just point at the process and you get everything, app and system-level.

DTrace is awesome.

**Comment by LowLevelMahn** [comment]
> One thing that's great about the Apple tools is that there's zero friction. No special recompiling to instrument the binary, just point at the process and you get everything, app and system-level.

but this will you get only sampling based profiling (with some sort of runtime patching needed) - or am im wrong?

for example the XRay-LLVM backend uses instrumentation for very exact tracing (i did not tested it so far, but want to)
https://www.youtube.com/watch?v=jyL-__zOGcU

**Comment by kklobe** [comment]
> > One thing that's great about the Apple tools is that there's zero friction. No special recompiling to instrument the binary, just point at the process and you get everything, app and system-level.
> 
> but this will you get only sampling based profiling (with some sort of runtime patching needed) - or am im wrong?

Due to kernel-level support via DTrace it's a bit more sophisticated than, say, userland sample-based profiling you might get on e.g. Windows with basic Intel VTune (although my last experience with it was over 15 years ago and a bit fuzzy).

> 
> for example the XRay-LLVM backend uses instrumentation for very exact tracing (i did not tested it so far, but want to)
> https://www.youtube.com/watch?v=jyL-__zOGcU

I get the impression that XRay is more targeted at a userland solution: http://llvm.1065342.n5.nabble.com/llvm-dev-RFC-XRay-A-Function-Call-Tracing-System-td94636.html

"dtrace: It requires kernel support to get going, as opposed to a purely userland solution. **Although what's being proposed is purely a subset of the functionality available to dtrace**, we find that the idea itself (function call tracing) has merit and would like to cite the existence of dtrace as "proof of validity" to the overall approach. :)"

It would be very interesting to compare the accuracy between the two, maybe we can get that going at some point.
